doc/development/guide/documentation.rst

```diff
@@ -1,39 +1,449 @@
+  examples include ``(array)``, ``(array[float])``, ``(dict)``, ``(dict[str, int])``, ``(Sequence[int] or int)``.
+  Note that square brakets are used to define the required types of container elements.
+
+  *Note:* if the argument has a default value, *do not* include it in the description if it is declared
+  in the signature. Sphinx will automatically extract the default and render it with the argument
+  description.
+
+* **Keyword Arguments:** To describe arbitrary keyword arguments provided via ``**kwargs``. Provide
+  the default values if relevant.
+
+* **Returns:/Yields:** To describe the return (yielded) value of the function (generator). This may
+  be omitted if the function signature/summary already sufficiently describe both the return value
+  and type.
+
+* **See also:** Provide optional links to relevant pages, functions, classes, methods, etc.
+
+* **Example:** To provide a minimal working example showing basic usage of the function. The example
+  should be *minimal* (reduce line counts where possible), but complete; a reader should be able to
+  copy and paste the example and get the same output. See :ref:`code_examples` for guidelines writing
+  useful code examples in docstrings.
+
+* **Usage details:** To provide a more complicated usage details showing different edge cases and
+  advanced usage of the function, as well as implementation details. This section of the docstring
+  is *collapsed by default* in the rendered Sphinx documentation to avoid overwhelming the reader with
+  information, however will be displayed in full to users in the terminal/Jupyter notebooks, so
+  do not let this section become too long.
+
+
+Classes
+~~~~~~~
+
+The class docstring is placed directly below the class definition:
+
+.. code-block:: python
+
+    class MyClass:
+        """Single sentence that summarizes the class.
+
+        Multi-line description of the class (optional, if required).
+
+        Args:
+            arg1 (type): Description.
+                Continuation line is indented if needed.
+            arg2 (type): description
+
+        Keyword Args:
+            kwarg1 (type): description
+
+        Attributes:
+            attr1 (type): description
+
+        Raises:
+            ExceptionType: description
+
+        .. seealso:: :func:`~.relevant_func`, :class:`~.RelevantClass` (optional)
+
+        **Example**
+
+        Minimal example with 1 or 2 code blocks (required).
+
+        .. UsageDetails::
+
+            More complicated use cases, options, and larger code blocks (optional).
+
+        **Related tutorials**
+
+        Links to any relevant PennyLane demos (optional).
+        """
+
+Docstrings for classes follow a similar structure as functions, with a few differences:
+
+* **Include initialization arguments.** The constructor ``__init__()`` is documented
+  here. Note that as the constructor simply returns an instance of the class, no ``Returns:``
+  section is included.
```

Comment:
I notice that sometimes people will use inline docstrings inside `__init__` to declare what each attribute does. What do we recommend in that case?